library/alloc/src/rc.rs

```diff
@@ -1628,7 +1628,7 @@ impl<T: ?Sized, A: Allocator> Rc<T, A> {
     pub unsafe fn get_mut_unchecked(this: &mut Self) -> &mut T {
         // We are careful to *not* create a reference covering the "count" fields, as
         // this would conflict with accesses to the reference counts (e.g. by `Weak`).
-        unsafe { &mut (*this.ptr.as_ptr()).value }
+        unsafe { &mut *ptr::addr_of_mut!((*this.ptr.as_ptr()).value) }
```

Comment:
I don't think this created any unnecessary references, the `addr_of_mut` roundtrip seems pointless.

Reply:
Maybe you're right, but this does result in different MIR output.
I thought that they were different because the previous line created a reference to the entire struct, not just that one field

Reply:
> I thought that they were different because the previous line created a reference to the entire struct, not just that one field

IIUC, `&mut (*ptr).field` does not create a reference to the entire struct. See also the following lines in Weak::inner.

https://github.com/rust-lang/rust/blob/87e1430893d55001034bd76c5dbe832d80bc50c3/library/alloc/src/sync.rs#L2849-L2852

Reply:
I saw that comment but I didn't quite trust it because it's apparently part of faulty code :) I guess I can close this tho